### DISCUSSION TITLE
Hydration Runtime Error

### DISCUSSION THREAD

**Discussion by tsimvris** [neuefische/questions#68] [Web] [answered]
I'm getting 3 Runtime Error in random times, specially when i am starting the Dev Server.

I've googled the error, and tried different solutions but nothing is working.

Here is the error ->

<img width="903" alt="Bildschirmfoto 2022-08-16 um 14 06 34" src="https://user-images.githubusercontent.com/101249132/184876973-23fe7d19-f570-4591-a01b-bc12e136a27b.png">

**Comment by PReichetanz** [comment]
Hi Petros,

it's hard to tell since you are not providing much information besides the error itself. Please commit and push your current state and add a link to your branch where the problem exists.

If you were already able to narrow down the problem to a specific file, please let us know as well.

You can check whether you have provided all necessary information [with the help of this question](https://github.com/neuefische/questions/discussions/11) asked by our coach Mareike.

**Comment by alexclaes** [comment, marked as answer]
I assume you are using the state management library `zustand` with the `persist` middleware. This is a common issue when working with Next.js (or any other server-side rendering logic).

The `persist` middleware stores and reads data in LocalStorage within the browser - which is not available on the server.

After the website is loaded from the server and displayed in the browser, the react rendering logic runs again (this is called hydration).

Since LocalStorage is not available on the server (only within the browser), the HTML generated on the server differs from the HTML generated in the browser. This is the reason for the error:

`Hydration failed because the initial UI does not match what was rendered on the server.`

There are multiple approaches to work around this issue. For Next.js applications you might use dynamic imports:

https://nextjs.org/docs/advanced-features/dynamic-import

Please make sure to set the option `{ ssr: false }`

https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr

Using the `suspense` option and wrapping the `<Suspense />` component is not required. This can be used to display a loading state.

Use dynamic import for all components, that access the `zustand` store. This tells Next.js to not render this component on the server, but only on the client to avoid this issue.

Once the dynamic import is used for a component, it also applies to all components nested within. Therefore, you only need to use dynamic imports on the upper most level in your component hierarchy (like on the Next.js page)
